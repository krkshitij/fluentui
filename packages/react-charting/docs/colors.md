# Colors

The color prop is currently optional for curves and lines in AreaChart and LineChart, respectively. If color is not provided, the chart elements will be rendered with colors from the default palette. The users can provide custom colors or use colors from these palettes:

## Qualitative Palette

![Qualitative color palette](colors/1.png)

This is the default color palette for data visualization. Currently, it has only 20 colors (including extension colors). We cycle through these colors sequentially to paint the chart elements without color.

To use colors from this palette:

```
import { QualitativePalette } from '@fluentui/react-charting';
```

Access colors like this:

```
color: QualitativePalette.dataVizSlot1,
```

## Semantic Palette

![Semantic color palette](colors/2.png)

To use colors from this palette:

```
import { SemanticPalette } from '@fluentui/react-charting';
```

Access colors like this:

```
color: SemanticPalette.info,
```
