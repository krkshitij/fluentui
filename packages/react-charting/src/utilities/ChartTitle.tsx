import * as React from 'react';
import { hasOverflow, ITooltipHostProps, TooltipHost, TooltipOverflowMode } from '@fluentui/react';
import { getAccessibleDataObject } from './index';
import { IAccessibilityProps } from '../types/index';

interface IChartTitleProps {
  className?: ITooltipHostProps['hostClassName'];
  accessibilityData?: IAccessibilityProps;
  children?: ITooltipHostProps['content'];
}

interface IChartTitleState {
  textOverflow: boolean;
}

export class ChartTitle extends React.Component<IChartTitleProps, IChartTitleState> {
  private _titleRef = React.createRef<HTMLSpanElement>();
  private _reqId: number;

  constructor(props: IChartTitleProps) {
    super(props);

    this.state = {
      textOverflow: false,
    };
  }

  public render(): React.ReactNode {
    const { className, accessibilityData, children } = this.props;

    return (
      <TooltipHost overflowMode={TooltipOverflowMode.Self} hostClassName={className} content={children}>
        <span
          ref={this._titleRef}
          {...getAccessibleDataObject(accessibilityData, 'text', this.state.textOverflow)}
          tabIndex={this.state.textOverflow ? 0 : -1}
        >
          {children}
        </span>
      </TooltipHost>
    );
  }

  public componentDidMount(): void {
    this._reqId = requestAnimationFrame(this._checkTextOverflow);
  }

  public componentWillUnmount(): void {
    cancelAnimationFrame(this._reqId);
  }

  private _checkTextOverflow = (): void => {
    const overflowElement = this._getTargetElement();
    const textOverflow = !!overflowElement && hasOverflow(overflowElement);
    if (textOverflow !== this.state.textOverflow) {
      this.setState({ textOverflow });
    }

    this._reqId = requestAnimationFrame(this._checkTextOverflow);
  };

  private _getTargetElement = (): HTMLElement | undefined => {
    if (!this._titleRef.current || !this._titleRef.current.parentElement) {
      return undefined;
    }

    return this._titleRef.current.parentElement;
  };
}
