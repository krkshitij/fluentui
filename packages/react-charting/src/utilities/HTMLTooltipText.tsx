import * as React from 'react';
import { hasOverflow, ITooltipHostProps, TooltipHost, TooltipOverflowMode } from '@fluentui/react';
import { getAccessibleDataObject } from './index';
import { IAccessibilityProps } from '../types/index';
import { Async } from '../Utilities';

interface IHTMLTooltipTextProps {
  className?: ITooltipHostProps['hostClassName'];
  content?: ITooltipHostProps['content'];
  accessibilityData?: IAccessibilityProps;
}

interface IHTMLTooltipTextState {
  textOverflow: boolean;
}

export class HTMLTooltipText extends React.Component<IHTMLTooltipTextProps, IHTMLTooltipTextState> {
  private _tooltipChild = React.createRef<HTMLSpanElement>();
  private _resizeObserver?: ResizeObserver;
  private _async: Async;

  constructor(props: IHTMLTooltipTextProps) {
    super(props);

    this.state = {
      textOverflow: false,
    };

    this._async = new Async(this);
  }

  public render(): React.ReactNode {
    const { className, content, accessibilityData } = this.props;

    return (
      <TooltipHost overflowMode={TooltipOverflowMode.Self} hostClassName={className} content={content}>
        <span
          {...getAccessibleDataObject(accessibilityData)}
          ref={this._tooltipChild}
          data-is-focusable={this.state.textOverflow}
          tabIndex={this.state.textOverflow ? 0 : -1}
        >
          {content}
        </span>
      </TooltipHost>
    );
  }

  public componentDidMount(): void {
    const overflowElement = this._getTargetElement();
    if (window.ResizeObserver && overflowElement) {
      this._resizeObserver = new ResizeObserver(this._async.debounce(this._checkTextOverflow, 500));
      this._resizeObserver.observe(overflowElement);
    }
  }

  public componentWillUnmount(): void {
    this._resizeObserver?.disconnect();
    this._async.dispose();
  }

  private _checkTextOverflow = (): void => {
    const overflowElement = this._getTargetElement();
    const textOverflow = !!overflowElement && hasOverflow(overflowElement);
    if (textOverflow !== this.state.textOverflow) {
      this.setState({ textOverflow });
    }
  };

  private _getTargetElement = (): HTMLElement | undefined => {
    if (!this._tooltipChild.current || !this._tooltipChild.current.parentElement) {
      return undefined;
    }

    return this._tooltipChild.current.parentElement;
  };
}
