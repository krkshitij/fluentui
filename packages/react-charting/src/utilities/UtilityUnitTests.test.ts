import * as utils from './utilities';
import * as colors from './colors';
import { TimeLocaleDefinition as d3TimeLocaleDefinition } from 'd3-time-format';

// Reference to the test plan: packages\react-charting\docs\TestPlans\Utilities\UnitTests.md

describe('Unit test to convert data to localized string', () => {
  test('Should return undefined when data provided is undefined', () => {
    expect(utils.convertToLocaleString(undefined)).toBeUndefined();
  });

  test('Should return the localised data in the given culture when input data is a string', () => {
    expect(utils.convertToLocaleString('text', 'en-GB')).toBe('text');
    expect(utils.convertToLocaleString('text', 'ar-SY')).toBe('text');
  });

  test('Should return the localised data in the given culture when the input data is a number', () => {
    expect(utils.convertToLocaleString(10, 'en-GB')).toBe('10');
    expect(utils.convertToLocaleString(2560, 'ar-SY')).toBe('٢٬٥٦٠');
  });

  test('Should return the localised data when the input data is a string containing a number', () => {
    expect(utils.convertToLocaleString('10', 'en-GB')).toBe('10');
    expect(utils.convertToLocaleString('1234', 'ar-SY')).toBe('١٬٢٣٤');
  });
});

describe('Unit test to return the accessible data object', () => {
  test('Should return the appropriate accessible data object no parameters are provided as input', () => {
    expect(utils.getAccessibleDataObject()).toEqual({
      role: 'text',
      'data-is-focusable': true,
      'aria-label': undefined,
      'aria-labelledby': undefined,
      'aria-describedby': undefined,
    });
  });

  test('Should return the appropriate accessible data object only role is provided as input', () => {
    expect(utils.getAccessibleDataObject(undefined, 'button')).toEqual({
      role: 'button',
      'data-is-focusable': true,
      'aria-label': undefined,
      'aria-labelledby': undefined,
      'aria-describedby': undefined,
    });
  });

  test('Should return the accessible data when both role and isDataFocusable is provided as input', () => {
    expect(utils.getAccessibleDataObject(undefined, 'text', false)).toEqual({
      role: 'text',
      'data-is-focusable': false,
      'aria-label': undefined,
      'aria-labelledby': undefined,
      'aria-describedby': undefined,
    });
  });

  test('Should return the appropriate accessible data object when all parameters are provided as input', () => {
    const accessibleData = {
      ariaLabel: 'Start button',
      ariaLabelledBy: 'Button',
      ariaDescribedBy: 'This is a start button',
    };
    expect(utils.getAccessibleDataObject(accessibleData, 'button', false)).toEqual({
      role: 'button',
      'data-is-focusable': false,
      'aria-label': 'Start button',
      'aria-labelledby': 'Button',
      'aria-describedby': 'This is a start button',
    });
  });
});

describe('Unit test for getting colors from token and returning the theme specific color', () => {
  test('Should return the token itself when the token is not from DataVizPallette', () => {
    expect(colors.getColorFromToken('blue')).toEqual('blue');
  });
  test('Should return the color code when the token is from DataVizPallette', () => {
    expect(colors.getColorFromToken('qualitative.1')).toEqual('#637cef');
  });

  test('Should return the first color when dark theme is disabled and length of colors list is more than 1', () => {
    expect(colors.getColorFromToken('qualitative.11', false)).toEqual('#3c51b4');
  });

  test('Should return the first color when dark theme is enabled and length of colors list is equal to 1', () => {
    expect(colors.getColorFromToken('qualitative.1', true)).toEqual('#637cef');
  });

  test('Should return the second color when dark theme is enabled and length of colors list is more than 1', () => {
    expect(colors.getColorFromToken('qualitative.11', true)).toEqual('#93a4f4');
  });
});

interface ICreateXAxisParams extends Partial<Omit<utils.IXAxisParams, 'domainNRangeValues'>> {
  domainNRangeValues?: Partial<utils.IDomainNRange>;
}
const createXAxisParams = (xAxisParams?: ICreateXAxisParams): utils.IXAxisParams => {
  const xAxisElement = document.createElementNS('http://www.w3.org/2000/svg', 'g');

  return {
    xAxisElement,
    containerHeight: 100,
    ...xAxisParams,
    domainNRangeValues: {
      dStartValue: 0,
      dEndValue: 100,
      rStartValue: 0,
      rEndValue: 100,
      ...xAxisParams?.domainNRangeValues,
    },
    margins: {
      left: 0,
      right: 0,
      top: 0,
      bottom: 0,
      ...xAxisParams?.margins,
    },
  };
};
const convertXAxisResultToJson = (
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  result: { xScale: any; tickValues: string[] },
  isStringAxis: boolean = false,
  tickCount: number = 6,
): [number, string][] => {
  const tickValues = isStringAxis ? result.tickValues : result.xScale.ticks(tickCount);
  return tickValues.map((item: number | Date | string, i: number) => [result.xScale(item), result.tickValues[i]]);
};
// eslint-disable-next-line @typescript-eslint/no-explicit-any
const matchResult = (result: any) => {
  expect(result).toMatchSnapshot();
};

describe('createNumericXAxis', () => {
  it('should render the x-axis labels correctly', () => {
    const xAxisParams = createXAxisParams();
    utils.createNumericXAxis(xAxisParams, utils.ChartTypes.VerticalBarChart);
    expect(xAxisParams.xAxisElement).toMatchSnapshot();
  });

  it('should create rounded x-axis labels when showRoundOffXTickValues is true', () => {
    const xAxisParams = createXAxisParams({
      domainNRangeValues: { dStartValue: 0.243, dEndValue: 0.433 },
      showRoundOffXTickValues: true,
    });
    const result = utils.createNumericXAxis(xAxisParams, utils.ChartTypes.VerticalBarChart);
    matchResult(convertXAxisResultToJson(result));
  });

  it('should render the x-axis labels correctly for specific tick size and padding values', () => {
    const xAxisParams = createXAxisParams({ xAxistickSize: 10, tickPadding: 5 });
    utils.createNumericXAxis(xAxisParams, utils.ChartTypes.VerticalBarChart);
    expect(xAxisParams.xAxisElement).toMatchSnapshot();
  });

  it('should create the x-axis labels correctly for a specific number of ticks', () => {
    const xAxisParams = createXAxisParams({ xAxisCount: 3 });
    const result = utils.createNumericXAxis(xAxisParams, utils.ChartTypes.VerticalBarChart);
    matchResult(convertXAxisResultToJson(result, false, xAxisParams.xAxisCount));
  });

  it('should render the x-axis labels correctly for horizontal bar chart with axis', () => {
    const xAxisParams = createXAxisParams();
    utils.createNumericXAxis(xAxisParams, utils.ChartTypes.HorizontalBarChartWithAxis);
    expect(xAxisParams.xAxisElement).toMatchSnapshot();
  });
});

describe('createDateXAxis', () => {
  const domainNRangeValues: ICreateXAxisParams['domainNRangeValues'] = {
    dStartValue: new Date(2021, 6, 1),
    dEndValue: new Date(2022, 5, 30),
  };

  it('should render the x-axis labels correctly', () => {
    const xAxisParams = createXAxisParams({ domainNRangeValues });
    utils.createDateXAxis(xAxisParams, {});
    expect(xAxisParams.xAxisElement).toMatchSnapshot();
  });

  it('should render the x-axis labels correctly for specific tick size and padding values', () => {
    const xAxisParams = createXAxisParams({ domainNRangeValues, xAxistickSize: 10, tickPadding: 5 });
    utils.createDateXAxis(xAxisParams, {});
    expect(xAxisParams.xAxisElement).toMatchSnapshot();
  });

  it('should create the x-axis labels correctly for a specific number of ticks', () => {
    const xAxisParams = createXAxisParams({ domainNRangeValues, xAxisCount: 12 });
    const result = utils.createDateXAxis(xAxisParams, {});
    matchResult(convertXAxisResultToJson(result, false, xAxisParams.xAxisCount));
  });

  it('should create the x-axis labels correctly when customDateTimeFormatter is provided', () => {
    const xAxisParams = createXAxisParams({ domainNRangeValues });
    const result = utils.createDateXAxis(xAxisParams, {}, undefined, undefined, undefined, (dateTime: Date) =>
      dateTime.toUTCString(),
    );
    matchResult(convertXAxisResultToJson(result));
  });

  it('should create the x-axis labels correctly when culture and options are provided', () => {
    const xAxisParams = createXAxisParams({ domainNRangeValues });
    const result = utils.createDateXAxis(xAxisParams, {}, 'ar-EG', { dateStyle: 'full', timeStyle: 'long' });
    matchResult(convertXAxisResultToJson(result));
  });

  it('should create the x-axis labels correctly when timeFormatLocale is provided', () => {
    // Fetched from https://unpkg.com/d3-time-format@2/locale/it-IT.json
    const timeFormatLocale: d3TimeLocaleDefinition = {
      dateTime: '%A %e %B %Y, %X',
      date: '%d/%m/%Y',
      time: '%H:%M:%S',
      periods: ['AM', 'PM'],
      days: ['Domenica', 'Lunedì', 'Martedì', 'Mercoledì', 'Giovedì', 'Venerdì', 'Sabato'],
      shortDays: ['Dom', 'Lun', 'Mar', 'Mer', 'Gio', 'Ven', 'Sab'],
      months: [
        'Gennaio',
        'Febbraio',
        'Marzo',
        'Aprile',
        'Maggio',
        'Giugno',
        'Luglio',
        'Agosto',
        'Settembre',
        'Ottobre',
        'Novembre',
        'Dicembre',
      ],
      shortMonths: ['Gen', 'Feb', 'Mar', 'Apr', 'Mag', 'Giu', 'Lug', 'Ago', 'Set', 'Ott', 'Nov', 'Dic'],
    };
    const xAxisParams = createXAxisParams({ domainNRangeValues });
    const result = utils.createDateXAxis(xAxisParams, {}, undefined, undefined, timeFormatLocale);
    matchResult(convertXAxisResultToJson(result));
  });

  it('should render the x-axis labels correctly when tickParams is provided', () => {
    const xAxisParams = createXAxisParams({ domainNRangeValues });
    utils.createDateXAxis(xAxisParams, {
      tickValues: [
        new Date(2021, 6, 1),
        new Date(2021, 9, 1),
        new Date(2022, 0, 1),
        new Date(2022, 3, 1),
        new Date(2022, 5, 30),
      ],
      tickFormat: '%a, %d %b %Y',
    });
    expect(xAxisParams.xAxisElement).toMatchSnapshot();
  });
});

describe('createStringXAxis', () => {
  const dataset: string[] = ['X-axis label 1', 'X-axis label 2', 'X-axis label 3'];

  it('should render the x-axis labels correctly', () => {
    const xAxisParams = createXAxisParams();
    utils.createStringXAxis(xAxisParams, {}, dataset);
    expect(xAxisParams.xAxisElement).toMatchSnapshot();
  });

  it('should render the x-axis labels correctly for specific tick size and padding values', () => {
    const xAxisParams = createXAxisParams({ xAxistickSize: 10, tickPadding: 5 });
    utils.createStringXAxis(xAxisParams, {}, dataset);
    expect(xAxisParams.xAxisElement).toMatchSnapshot();
  });

  it('should create the x-axis labels correctly for specific padding values', () => {
    const xAxisParams = createXAxisParams({ xAxisPadding: 0.5 });
    const result = utils.createStringXAxis(xAxisParams, {}, dataset);
    matchResult(convertXAxisResultToJson(result, true));
  });

  it('should create the x-axis labels correctly for specific inner and outer padding values', () => {
    const xAxisParams = createXAxisParams({ xAxisInnerPadding: 0.5, xAxisOuterPadding: 1 / 3 });
    const result = utils.createStringXAxis(xAxisParams, {}, dataset);
    matchResult(convertXAxisResultToJson(result, true));
  });
});

test('prepareDatapoints should return an array of uniformly distributed data points', () => {
  const result = utils.prepareDatapoints(100, 0, 3);
  matchResult(result);
});
