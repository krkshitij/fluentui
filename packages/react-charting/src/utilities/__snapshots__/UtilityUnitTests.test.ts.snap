// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createDateXAxis should create the x-axis labels correctly for a specific number of ticks 1`] = `
Array [
  Array [
    0,
    "July",
  ],
  Array [
    8.516483516483516,
    "August",
  ],
  Array [
    17.032967032967033,
    "September",
  ],
  Array [
    25.274725274725274,
    "October",
  ],
  Array [
    33.791208791208796,
    "November",
  ],
  Array [
    42.032967032967036,
    "December",
  ],
  Array [
    50.54945054945055,
    "2022",
  ],
  Array [
    59.065934065934066,
    "February",
  ],
  Array [
    66.75824175824175,
    "March",
  ],
  Array [
    75.27472527472527,
    "April",
  ],
  Array [
    83.51648351648352,
    "May",
  ],
  Array [
    92.03296703296702,
    "June",
  ],
]
`;

exports[`createDateXAxis should create the x-axis labels correctly when culture and options are provided 1`] = `
Array [
  Array [
    0,
    "الخميس، ١ يوليو ٢٠٢١ في ١٢:٠٠:٠٠ ص غرينتش+٥:٣٠",
  ],
  Array [
    25.274725274725274,
    "الجمعة، ١ أكتوبر ٢٠٢١ في ١٢:٠٠:٠٠ ص غرينتش+٥:٣٠",
  ],
  Array [
    50.54945054945055,
    "السبت، ١ يناير ٢٠٢٢ في ١٢:٠٠:٠٠ ص غرينتش+٥:٣٠",
  ],
  Array [
    75.27472527472527,
    "الجمعة، ١ أبريل ٢٠٢٢ في ١٢:٠٠:٠٠ ص غرينتش+٥:٣٠",
  ],
]
`;

exports[`createDateXAxis should create the x-axis labels correctly when customDateTimeFormatter is provided 1`] = `
Array [
  Array [
    0,
    "Wed, 30 Jun 2021 18:30:00 GMT",
  ],
  Array [
    25.274725274725274,
    "Thu, 30 Sep 2021 18:30:00 GMT",
  ],
  Array [
    50.54945054945055,
    "Fri, 31 Dec 2021 18:30:00 GMT",
  ],
  Array [
    75.27472527472527,
    "Thu, 31 Mar 2022 18:30:00 GMT",
  ],
]
`;

exports[`createDateXAxis should create the x-axis labels correctly when timeFormatLocale is provided 1`] = `
Array [
  Array [
    0,
    "Luglio",
  ],
  Array [
    25.274725274725274,
    "Ottobre",
  ],
  Array [
    50.54945054945055,
    "2022",
  ],
  Array [
    75.27472527472527,
    "Aprile",
  ],
]
`;

exports[`createDateXAxis should render the x-axis labels correctly 1`] = `
<g
  fill="none"
  font-family="sans-serif"
  font-size="10"
  text-anchor="middle"
>
  <path
    class="domain"
    d="M0.5,6V0.5H100.5V6"
    stroke="#000"
  />
  <g
    class="tick"
    opacity="1"
    transform="translate(0.5,0)"
  >
    <line
      stroke="#000"
      y2="6"
    />
    <text
      aria-hidden="true"
      dy="0.71em"
      fill="#000"
      y="12"
    >
      July
    </text>
  </g>
  <g
    class="tick"
    opacity="1"
    transform="translate(25.774725274725274,0)"
  >
    <line
      stroke="#000"
      y2="6"
    />
    <text
      aria-hidden="true"
      dy="0.71em"
      fill="#000"
      y="12"
    >
      October
    </text>
  </g>
  <g
    class="tick"
    opacity="1"
    transform="translate(51.04945054945055,0)"
  >
    <line
      stroke="#000"
      y2="6"
    />
    <text
      aria-hidden="true"
      dy="0.71em"
      fill="#000"
      y="12"
    >
      2022
    </text>
  </g>
  <g
    class="tick"
    opacity="1"
    transform="translate(75.77472527472527,0)"
  >
    <line
      stroke="#000"
      y2="6"
    />
    <text
      aria-hidden="true"
      dy="0.71em"
      fill="#000"
      y="12"
    >
      April
    </text>
  </g>
</g>
`;

exports[`createDateXAxis should render the x-axis labels correctly for specific tick size and padding values 1`] = `
<g
  fill="none"
  font-family="sans-serif"
  font-size="10"
  text-anchor="middle"
>
  <path
    class="domain"
    d="M0.5,10V0.5H100.5V10"
    stroke="#000"
  />
  <g
    class="tick"
    opacity="1"
    transform="translate(0.5,0)"
  >
    <line
      stroke="#000"
      y2="10"
    />
    <text
      aria-hidden="true"
      dy="0.71em"
      fill="#000"
      y="15"
    >
      July
    </text>
  </g>
  <g
    class="tick"
    opacity="1"
    transform="translate(25.774725274725274,0)"
  >
    <line
      stroke="#000"
      y2="10"
    />
    <text
      aria-hidden="true"
      dy="0.71em"
      fill="#000"
      y="15"
    >
      October
    </text>
  </g>
  <g
    class="tick"
    opacity="1"
    transform="translate(51.04945054945055,0)"
  >
    <line
      stroke="#000"
      y2="10"
    />
    <text
      aria-hidden="true"
      dy="0.71em"
      fill="#000"
      y="15"
    >
      2022
    </text>
  </g>
  <g
    class="tick"
    opacity="1"
    transform="translate(75.77472527472527,0)"
  >
    <line
      stroke="#000"
      y2="10"
    />
    <text
      aria-hidden="true"
      dy="0.71em"
      fill="#000"
      y="15"
    >
      April
    </text>
  </g>
</g>
`;

exports[`createDateXAxis should render the x-axis labels correctly when tickParams is provided 1`] = `
<g
  fill="none"
  font-family="sans-serif"
  font-size="10"
  text-anchor="middle"
>
  <path
    class="domain"
    d="M0.5,6V0.5H100.5V6"
    stroke="#000"
  />
  <g
    class="tick"
    opacity="1"
    transform="translate(0.5,0)"
  >
    <line
      stroke="#000"
      y2="6"
    />
    <text
      aria-hidden="true"
      dy="0.71em"
      fill="#000"
      y="12"
    >
      Thu, 01 Jul 2021
    </text>
  </g>
  <g
    class="tick"
    opacity="1"
    transform="translate(25.774725274725274,0)"
  >
    <line
      stroke="#000"
      y2="6"
    />
    <text
      aria-hidden="true"
      dy="0.71em"
      fill="#000"
      y="12"
    >
      Fri, 01 Oct 2021
    </text>
  </g>
  <g
    class="tick"
    opacity="1"
    transform="translate(51.04945054945055,0)"
  >
    <line
      stroke="#000"
      y2="6"
    />
    <text
      aria-hidden="true"
      dy="0.71em"
      fill="#000"
      y="12"
    >
      Sat, 01 Jan 2022
    </text>
  </g>
  <g
    class="tick"
    opacity="1"
    transform="translate(75.77472527472527,0)"
  >
    <line
      stroke="#000"
      y2="6"
    />
    <text
      aria-hidden="true"
      dy="0.71em"
      fill="#000"
      y="12"
    >
      Fri, 01 Apr 2022
    </text>
  </g>
  <g
    class="tick"
    opacity="1"
    transform="translate(100.5,0)"
  >
    <line
      stroke="#000"
      y2="6"
    />
    <text
      aria-hidden="true"
      dy="0.71em"
      fill="#000"
      y="12"
    >
      Thu, 30 Jun 2022
    </text>
  </g>
</g>
`;

exports[`createNumericXAxis should create rounded x-axis labels when showRoundOffXTickValues is true 1`] = `
Array [
  Array [
    5.000000000000004,
    "0.25",
  ],
  Array [
    30,
    "0.3",
  ],
  Array [
    54.99999999999999,
    "0.35",
  ],
  Array [
    80.00000000000001,
    "0.4",
  ],
]
`;

exports[`createNumericXAxis should create the x-axis labels correctly for a specific number of ticks 1`] = `
Array [
  Array [
    0,
    0,
  ],
  Array [
    50,
    "50",
  ],
  Array [
    100,
    "100",
  ],
]
`;

exports[`createNumericXAxis should render the x-axis labels correctly 1`] = `
<g
  fill="none"
  font-family="sans-serif"
  font-size="10"
  text-anchor="middle"
>
  <path
    class="domain"
    d="M0.5,6V0.5H100.5V6"
    stroke="#000"
  />
  <g
    class="tick"
    opacity="1"
    transform="translate(0.5,0)"
  >
    <line
      stroke="#000"
      y2="6"
    />
    <text
      aria-hidden="true"
      dy="0.71em"
      fill="#000"
      y="16"
    >
      0
    </text>
  </g>
  <g
    class="tick"
    opacity="1"
    transform="translate(20.5,0)"
  >
    <line
      stroke="#000"
      y2="6"
    />
    <text
      aria-hidden="true"
      dy="0.71em"
      fill="#000"
      y="16"
    >
      20
    </text>
  </g>
  <g
    class="tick"
    opacity="1"
    transform="translate(40.5,0)"
  >
    <line
      stroke="#000"
      y2="6"
    />
    <text
      aria-hidden="true"
      dy="0.71em"
      fill="#000"
      y="16"
    >
      40
    </text>
  </g>
  <g
    class="tick"
    opacity="1"
    transform="translate(60.5,0)"
  >
    <line
      stroke="#000"
      y2="6"
    />
    <text
      aria-hidden="true"
      dy="0.71em"
      fill="#000"
      y="16"
    >
      60
    </text>
  </g>
  <g
    class="tick"
    opacity="1"
    transform="translate(80.5,0)"
  >
    <line
      stroke="#000"
      y2="6"
    />
    <text
      aria-hidden="true"
      dy="0.71em"
      fill="#000"
      y="16"
    >
      80
    </text>
  </g>
  <g
    class="tick"
    opacity="1"
    transform="translate(100.5,0)"
  >
    <line
      stroke="#000"
      y2="6"
    />
    <text
      aria-hidden="true"
      dy="0.71em"
      fill="#000"
      y="16"
    >
      100
    </text>
  </g>
</g>
`;

exports[`createNumericXAxis should render the x-axis labels correctly for horizontal bar chart with axis 1`] = `
<g
  fill="none"
  font-family="sans-serif"
  font-size="10"
  text-anchor="middle"
>
  <path
    class="domain"
    d="M0.5,6V0.5H100.5V6"
    stroke="#000"
  />
  <g
    class="tick"
    opacity="1"
    transform="translate(0.5,0)"
  >
    <line
      stroke="#000"
      y2="-100"
    />
    <text
      aria-hidden="true"
      dy="0.71em"
      fill="#000"
      y="10"
    >
      0
    </text>
  </g>
  <g
    class="tick"
    opacity="1"
    transform="translate(20.5,0)"
  >
    <line
      stroke="#000"
      y2="-100"
    />
    <text
      aria-hidden="true"
      dy="0.71em"
      fill="#000"
      y="10"
    >
      20
    </text>
  </g>
  <g
    class="tick"
    opacity="1"
    transform="translate(40.5,0)"
  >
    <line
      stroke="#000"
      y2="-100"
    />
    <text
      aria-hidden="true"
      dy="0.71em"
      fill="#000"
      y="10"
    >
      40
    </text>
  </g>
  <g
    class="tick"
    opacity="1"
    transform="translate(60.5,0)"
  >
    <line
      stroke="#000"
      y2="-100"
    />
    <text
      aria-hidden="true"
      dy="0.71em"
      fill="#000"
      y="10"
    >
      60
    </text>
  </g>
  <g
    class="tick"
    opacity="1"
    transform="translate(80.5,0)"
  >
    <line
      stroke="#000"
      y2="-100"
    />
    <text
      aria-hidden="true"
      dy="0.71em"
      fill="#000"
      y="10"
    >
      80
    </text>
  </g>
  <g
    class="tick"
    opacity="1"
    transform="translate(100.5,0)"
  >
    <line
      stroke="#000"
      y2="-100"
    />
    <text
      aria-hidden="true"
      dy="0.71em"
      fill="#000"
      y="10"
    >
      100
    </text>
  </g>
</g>
`;

exports[`createNumericXAxis should render the x-axis labels correctly for specific tick size and padding values 1`] = `
<g
  fill="none"
  font-family="sans-serif"
  font-size="10"
  text-anchor="middle"
>
  <path
    class="domain"
    d="M0.5,10V0.5H100.5V10"
    stroke="#000"
  />
  <g
    class="tick"
    opacity="1"
    transform="translate(0.5,0)"
  >
    <line
      stroke="#000"
      y2="10"
    />
    <text
      aria-hidden="true"
      dy="0.71em"
      fill="#000"
      y="15"
    >
      0
    </text>
  </g>
  <g
    class="tick"
    opacity="1"
    transform="translate(20.5,0)"
  >
    <line
      stroke="#000"
      y2="10"
    />
    <text
      aria-hidden="true"
      dy="0.71em"
      fill="#000"
      y="15"
    >
      20
    </text>
  </g>
  <g
    class="tick"
    opacity="1"
    transform="translate(40.5,0)"
  >
    <line
      stroke="#000"
      y2="10"
    />
    <text
      aria-hidden="true"
      dy="0.71em"
      fill="#000"
      y="15"
    >
      40
    </text>
  </g>
  <g
    class="tick"
    opacity="1"
    transform="translate(60.5,0)"
  >
    <line
      stroke="#000"
      y2="10"
    />
    <text
      aria-hidden="true"
      dy="0.71em"
      fill="#000"
      y="15"
    >
      60
    </text>
  </g>
  <g
    class="tick"
    opacity="1"
    transform="translate(80.5,0)"
  >
    <line
      stroke="#000"
      y2="10"
    />
    <text
      aria-hidden="true"
      dy="0.71em"
      fill="#000"
      y="15"
    >
      80
    </text>
  </g>
  <g
    class="tick"
    opacity="1"
    transform="translate(100.5,0)"
  >
    <line
      stroke="#000"
      y2="10"
    />
    <text
      aria-hidden="true"
      dy="0.71em"
      fill="#000"
      y="15"
    >
      100
    </text>
  </g>
</g>
`;

exports[`createStringXAxis should create the x-axis labels correctly for specific inner and outer padding values 1`] = `
Array [
  Array [
    10.526315789473685,
    "X-axis label 1",
  ],
  Array [
    42.10526315789474,
    "X-axis label 2",
  ],
  Array [
    73.6842105263158,
    "X-axis label 3",
  ],
]
`;

exports[`createStringXAxis should create the x-axis labels correctly for specific padding values 1`] = `
Array [
  Array [
    14.285714285714285,
    "X-axis label 1",
  ],
  Array [
    42.85714285714286,
    "X-axis label 2",
  ],
  Array [
    71.42857142857143,
    "X-axis label 3",
  ],
]
`;

exports[`createStringXAxis should render the x-axis labels correctly 1`] = `
<g
  fill="none"
  font-family="sans-serif"
  font-size="10"
  text-anchor="middle"
>
  <path
    class="domain"
    d="M0.5,6V0.5H100.5V6"
    stroke="#000"
  />
  <g
    class="tick"
    opacity="1"
    transform="translate(17.741935483870968,0)"
  >
    <line
      stroke="#000"
      y2="6"
    />
    <text
      aria-hidden="true"
      dy="0.71em"
      fill="#000"
      y="16"
    >
      X-axis label 1
    </text>
  </g>
  <g
    class="tick"
    opacity="1"
    transform="translate(50,0)"
  >
    <line
      stroke="#000"
      y2="6"
    />
    <text
      aria-hidden="true"
      dy="0.71em"
      fill="#000"
      y="16"
    >
      X-axis label 2
    </text>
  </g>
  <g
    class="tick"
    opacity="1"
    transform="translate(82.25806451612904,0)"
  >
    <line
      stroke="#000"
      y2="6"
    />
    <text
      aria-hidden="true"
      dy="0.71em"
      fill="#000"
      y="16"
    >
      X-axis label 3
    </text>
  </g>
</g>
`;

exports[`createStringXAxis should render the x-axis labels correctly for specific tick size and padding values 1`] = `
<g
  fill="none"
  font-family="sans-serif"
  font-size="10"
  text-anchor="middle"
>
  <path
    class="domain"
    d="M0.5,10V0.5H100.5V10"
    stroke="#000"
  />
  <g
    class="tick"
    opacity="1"
    transform="translate(17.741935483870968,0)"
  >
    <line
      stroke="#000"
      y2="10"
    />
    <text
      aria-hidden="true"
      dy="0.71em"
      fill="#000"
      y="15"
    >
      X-axis label 1
    </text>
  </g>
  <g
    class="tick"
    opacity="1"
    transform="translate(50,0)"
  >
    <line
      stroke="#000"
      y2="10"
    />
    <text
      aria-hidden="true"
      dy="0.71em"
      fill="#000"
      y="15"
    >
      X-axis label 2
    </text>
  </g>
  <g
    class="tick"
    opacity="1"
    transform="translate(82.25806451612904,0)"
  >
    <line
      stroke="#000"
      y2="10"
    />
    <text
      aria-hidden="true"
      dy="0.71em"
      fill="#000"
      y="15"
    >
      X-axis label 3
    </text>
  </g>
</g>
`;

exports[`prepareDatapoints should return an array of uniformly distributed data points 1`] = `
Array [
  0,
  34,
  68,
  102,
]
`;
